File: Design1_Problem1.py
# Time Complexity : O(1) for put, get, remove
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : yes
# Any problem you faced while coding this : None

# Your code here along with comments explaining your approach : Took a 2D array to increase the efficiency and stored the boolean values for the value existance

#2D array
class MyHashSet(object):
    
    def __init__(self):
        # self.hashset = bytearray(1000001)
        self.hashset = [[] for i in range(1000)]
        self.bucket = 1000
        self.bucket2 = 1000

    def add(self, key):
        """
        :type key: int
        :rtype: None
        """
        hash1 = key%self.bucket
        if not self.hashset[hash1]:
            if hash1 == 0:
                self.hashset[hash1] = [0]*1001
            else:
                self.hashset[hash1] = [0]*1000
        hash2 = key//self.bucket2
        self.hashset[hash1][hash2] = 1
    def remove(self, key):
        """
        :type key: int
        :rtype: None
        """
        hash1 = key%self.bucket
        hash2 = key//self.bucket2
        if self.hashset[hash1]:
            self.hashset[hash1][hash2] = 0
        # if self.hashset[key]:
        #     self.hashset[key] = 0

    def contains(self, key):
        """
        :type key: int
        :rtype: bool
        """
        hash1 = key%self.bucket
        hash2 = key//self.bucket2
        return False if not self.hashset[hash1] else self.hashset[hash1][hash2]

#1D Array
# class MyHashSet(object):

#     def __init__(self):
#         self.hashset = bytearray(1000001)

#     def add(self, key):
#         """
#         :type key: int
#         :rtype: None
#         """
#         if not self.hashset[key]:
#             self.hashset[key] = 1

#     def remove(self, key):
#         """
#         :type key: int
#         :rtype: None
#         """
#         if self.hashset[key]:
#             self.hashset[key] = 0

#     def contains(self, key):
#         """
#         :type key: int
#         :rtype: bool
#         """
#         return self.hashset[key]


# Your MyHashSet object will be instantiated and called as such:
# obj = MyHashSet()
# obj.add(key)
# obj.remove(key)
# param_3 = obj.contains(key)

File: Design1_Problem1.java
// Time Complexity : O(1) FOR PUT, GET, DELETE in BEST CASE (If only 1 element in the list or that index is null) 
  //O(N) for PUT, GET, DELETE in Average case, WORST CASE. 

// Space Complexity : O(10^6)
// Did this code successfully run on Leetcode : Yes

// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
/*

Created primary array of Type Nodes with size 10000 --> Each array will be having max 100 items of key-value pairs
Used hasing for calculating the index to store key-value pair

Used chaining (LINKEDLIST)to add, get and delete key-value pair in hashmap

Implemented find() to get the current node in list stored at specific index for given key

If the key is not in the list --> Will add new node in the list after prev node returned from find() 
If the key is already there, just update the value for that key


Same for delete(key-value pair) and get value for any key : Used find() to get the current node in list stored at specific index for given key

*/



class MyHashMap {
    class Node{
        int key;
        int val;
        Node next;
        
        public Node(int key, int val){
            this.key = key;
            this.val = val;
            this.next = null;
        }
    }
    private Node[] nodes;
    private int buckets;
    public MyHashMap() {
        this.buckets = 10000;
        nodes = new Node[buckets];
    }
    private int hash(int key){
        return key % buckets;
    }
    
    private Node find(Node head, int key){
        Node prev = null;
        Node curr = head;
        
        while(curr != null && curr.key != key){
            prev = curr;
            curr = curr.next;
        }
        
        return prev;
    }
    public void put(int key, int value) {
        int h = hash(key);
        
        if(nodes[h] == null){
            nodes[h] = new Node(-1,-1);      
        }
        Node prev = find(nodes[h], key);
        
        if(prev.next == null){
            prev.next = new Node(key,value);    
        }
        else{
            prev.next.val = value;
        }
    }
    
    public int get(int key) {
        int h = hash(key);
        Node prev = find(nodes[h], key);
        
        if(nodes[h] == null){
            return -1;   
        }
        
        if(prev.next == null)    return -1;
        
        return prev.next.val;
    }
    
    public void remove(int key) {
        int h = hash(key);
        Node prev = find(nodes[h], key);
        
        if(nodes[h] == null){
            return;   
        }
        if(prev.next == null) return;
        
        prev.next = prev.next.next;
    }
}




public class Design1_Problem1{
    public static void main(String[] args){
        MyHashMap map = new MyHashMap();

        map.put(1,1); map.put(2,2);
        System.out.println("map.get(1) : " + map.get(1) );
        System.out.println("map.get(3) : " + map.get(3) );
        
        map.put(2,1);

        System.out.println("map.get(2) : " + map.get(2) );
        map.remove(2);

        System.out.println("map.get(2) : " + map.get(2) );

    }
}